pennylane_lightning/core/src/simulators/lightning_tensor/cutn/CutnBase.hpp

```diff
@@ -0,0 +1,222 @@
+    };
+
+    /**
+     * @brief Get device and Cuda stream information (device ID and the
+     * associated Cuda stream ID).
+     *
+     * @return DevTag
+     */
+    [[nodiscard]] auto getDevTag() -> DevTag<int> & { return dev_tag_; }
+
+    /**
+     * @brief Get the memory size used for a work space
+     *
+     * @return size_t
+     */
+    size_t getWorkSpaceMemorySize(cutensornetWorkspaceDescriptor_t &workDesc) {
+        int64_t worksize{0};
+
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetWorkspaceGetMemorySize(
+            /* const cutensornetHandle_t */ getCutnHandle(),
+            /* cutensornetWorkspaceDescriptor_t */ workDesc,
+            /* cutensornetWorksizePref_t */
+            CUTENSORNET_WORKSIZE_PREF_RECOMMENDED,
+            /* cutensornetMemspace_t*/ CUTENSORNET_MEMSPACE_DEVICE,
+            /* cutensornetWorkspaceKind_t */ CUTENSORNET_WORKSPACE_SCRATCH,
+            /*  int64_t * */ &worksize));
+
+        return worksize;
+    }
+
+    /**
+     * @brief Set the memory for a work space
+     *
+     * @param workDesc cutensornet work space descriptor
+     * @param scratchPtr Pointer to scratch memory
+     * @param worksize Memory size of a work space
+     */
+    void setWorkSpaceMemory(cutensornetWorkspaceDescriptor_t &workDesc,
+                            void *scratchPtr, int64_t &worksize) {
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetWorkspaceSetMemory(
+            /* const cutensornetHandle_t */ getCutnHandle(),
+            /* cutensornetWorkspaceDescriptor_t */ workDesc,
+            /* cutensornetMemspace_t*/ CUTENSORNET_MEMSPACE_DEVICE,
+            /* cutensornetWorkspaceKind_t */ CUTENSORNET_WORKSPACE_SCRATCH,
+            /* void *const */ scratchPtr,
+            /* int64_t */ worksize));
+    }
+
+    /**
+     * @brief Save quantumState information to data provided by a user
+     *
+     * @param extentsPtr Extents of each sites
+     * @param tensorPtr Pointer to tensors provided by a user
+     */
+    void computeState(std::vector<int64_t *> &extentsPtr,
+                      std::vector<void *> &tensorPtr) {
+        cutensornetWorkspaceDescriptor_t workDesc;
+        PL_CUTENSORNET_IS_SUCCESS(
+            cutensornetCreateWorkspaceDescriptor(getCutnHandle(), &workDesc));
+
+        // TODO we assign half (magic number is) of free memory size to the
+        // maximum memory usage.
+        const std::size_t scratchSize = cuUtil::getFreeMemorySize() / 2;
+
+        PL_CUTENSORNET_IS_SUCCESS(cutensornetStatePrepare(
+            /* const cutensornetHandle_t */ getCutnHandle(),
+            /* cutensornetState_t */ getQuantumState(),
+            /* size_t maxWorkspaceSizeDevice */ scratchSize,
+            /* cutensornetWorkspaceDescriptor_t */ workDesc,
+            /*  cudaStream_t unused in v24.03*/ 0x0));
+
+        int64_t worksize = getWorkSpaceMemorySize(workDesc);
+
+        // Ensure data is aligned by 256 bytes
+        worksize += int64_t{256} - worksize % int64_t{256};
```

Comment:
Why is this needed?

Reply:
Good question! This line ensures the `workDesc` to be `256-byte` aligned as required [here](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/functions.html#_CPPv423cutensornetStateComputeK19cutensornetHandle_t18cutensornetState_t32cutensornetWorkspaceDescriptor_tA_P7int64_tA_P7int64_tA_Pv12cudaStream_t)